Add --input-dir and tesseract-input/ mount point

angela-ko opened the pull request:
#### Description of changes
- Added `--input-dir` to tesseract run cli commands. 
- If `--input-dir` is specified, instead of a json payload, we would be expecting a relative path to the --input-dir and this would be located under `/tesseract-input`

I debated between putting this flag in runtime vs sdk, but I decided to go with sdk since runtime shouldn't need to know anything about the paths of stuff that lives outside of it (the container). This might make the argument a bit confusing since `--input-dir` would need to be before `apply`, e.g. `tesseract run --input-dir helloworld apply inputs.json`. 

I also swapped to the name `input-dir` instead of `path` since there already exists flags for runtime names `input-paths` and `output-paths`, and since we are expecting a directory, this makes a bit more sense? 

#### Testing done
Unit testing


- [ ] Additional Testing to be done
- [ ] More checking of `--input-dir` path vs `json_payload`

angela-ko commented:
Did some changes and integrated more with the existing @ so it makes more sense, mainly because of implementation problems. 

- For the PAYLOAD arg in Tesseract runtime, we dynamically create the commands and then use a callback function to trigger the actual functionalities. However, the callback function even with the context cannot know future params and their values (also arguments get parsed first before options), thus there was no way to achieve the previous implementation of `if input-path is specified, do this with the payload argument` instead. This also made specifying the `input-path` as a flag in the runtime pretty useless since by the time we parse `input-path`, PAYLOAD would already be processed. 
- Since we mentioned usage of `env var` on slack, this seemed like a good time to use `env var`. 
- However, if the env var is set even if no `--input-path` is specified on the Tesseract core side, **there needs to be a way to identify that what we are specifying is a path not an actual json PAYLOAD.. if only there was an easy way to identify.. oh wait there is. The `@` logic.** I decided to utilize the existing `@` logic since we use `@` if we want to specify a filepath instead of an actual JSON PAYLOAD. The only modification is we would use the env var path and the `@` would be relative to that env var path. That way, it is easier to distinguish between a JSON Payload and a filepath. 


So now, we just need to set the env var `TESSERACT_INPUT_PATH` if we want to use Tesseract runtime directly and the previous method of using `@` for a filepath would still be relevant in this case as well, except it would be a relative path to the env var.

angela-ko commented on tests/sdk_tests/test_engine.py:
What is the command you're running?

`@{infile}` should be relative to the input path, yes. I need to fix the unit test. This is what happens when you code on a red-eye flight.. I was wondering why I had separate test cases before..

angela-ko replied:
This should've been handled in runtime though, where we resolve the paths... Let me look into this

```
if value.startswith("@"):
        value = value[1:]
        tesseract_input_path = os.environ.get("TESSERACT_INPUT_PATH")
        if tesseract_input_path:
            if not os.access(tesseract_input_path, os.R_OK):
                raise click.BadParameter(
                    f"Tesseract input path {tesseract_input_path} is not accessible."
                )
            value = os.path.join(tesseract_input_path, value)
            typer.echo(
                f"Resolving @path to {value}. \n"
                "Clear the TESSERACT_INPUT_PATH environment variable to disable this behavior."
            )
        base_dir = Path(value).parent
```

angela-ko replied:
Ah I didn't catch my mistake in my tests because `os.path.join` can join together two absolute paths without problems. Therefore it doesn't always have to be a relative path, as absolute paths (subdirectory of the input path) would still work. 

Also path parsing would have to be in the runtime cli and not in engine if we wish for the logic to work when people run `tesseract_runtime` directly, while there is a minor bug in the code chunk you listed (good catch!) the major problem is because the environment variable we are setting isn't getting passed into the runtime side.... So I think I have to change something with the config.

angela-ko replied:
I think it's fixed now ... Can't seem to get tesseract-runtime to update locally but based on ci it seems to be working?

angela-ko commented:
It is not fixed. 

I am not sure how config works and how to pass the environment variable to the runtime side... Do I need to update the `sdk/config.py` as well? But that one only has docker related stuff so I'm guessing no.. How exactly is the config passed to the runtime? I tried some stuff out but it seems that config only affects DockerFile aka tesseract build... 

so maybe I need #250 to pass the env var.

angela-ko commented on tesseract_core/sdk/templates/Dockerfile.base:
```diff
@@ -74,7 +74,8 @@ USER tesseractor
     TESSERACT_DESCRIPTION="{{ config.description | replace('"', '\\"') | replace('\n', '\\n') }}" \
-    TESSERACT_API_PATH="/tesseract/tesseract_api.py"
+    TESSERACT_API_PATH="/tesseract/tesseract_api.py" \
+    TESSERACT_INPUT_PATH="{{ config.input_path | replace('"', '\\"') | replace('\n', '\\n') }}"
```
Oh yeah this needs to be reverted - was a part of trying to fix the env variable thing but i think we just need #250

angela-ko commented on tesseract_core/runtime/cli.py:
```diff
@@ -82,12 +82,24 @@ def _parse_arg_callback(
     if value.startswith("@"):
-        base_dir = Path(value[1:]).parent
-        value_format = guess_format_from_path(value[1:])
+        value = value[1:]
+        tesseract_input_path = get_config().input_path
+        if tesseract_input_path:
+            if not os.access(tesseract_input_path, os.R_OK):
+                raise click.BadParameter(
+                    f"Tesseract input path {tesseract_input_path} is not accessible."
+                )
+            value = os.path.join(tesseract_input_path, value)
+            typer.echo(
+                f"Resolving @path to {value}. \n"
+                "Clear the TESSERACT_INPUT_PATH environment variable to disable this behavior."
+            )
+        base_dir = Path(value).parent
+        value_format = guess_format_from_path(value)
```
They do seem like they would be coupled though. Otherwise how would the interaction work? `--input-path` affects what we pass in for the PAYLOAD argument because otherwise it would only create a input path that we need to pull inputs from but the actual inputs still need to be passed in somehow (through PAYLOAD).

angela-ko replied:
How would we test it then from cli without reading the filepath? It still doesn't feel like a very intrusive change to @ and seems natural that after we mount `--input-path` to also want the filepath to be read from there as well.

dionhaefner commented on tesseract_core/sdk/cli.py:
```diff
@@ -847,6 +847,14 @@ def run_container(
             help=("User to run the Tesseract as e.g. '1000' or '1000:1000' (uid:gid)."),
         ),
     ] = None,
+    input_path: Annotated[
+        str | None,
+        typer.Option(
+            "--input-path",
+            help="Input path to read input files from, such as local directory or S3 URI "
+            "(may be anything supported by fsspec).",
+        ),
+    ] = None,
```
Should we pass `hidden=True` to `typer.Option` then? That way it's available for testing but not exposed to users (yet).

angela-ko commented:
```
(tesseract) 柯 ~/work/tesseract-core » tesseract run --help
                                                                                                                      
 Usage: tesseract run [OPTIONS] TESSERACT_IMAGE [CMD] [ARGS]...                                                       
                                                                                                                      
 Execute a command in a Tesseract.                                                                                    
 This command starts a Tesseract instance and executes the given command.                                             
                                                                                                                      
╭─ Arguments ────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ *    tesseract_image      TEXT       Tesseract image name [default: None] [required]                               │
│      cmd                  [CMD]      Tesseract command to run                                                      │
│      args                 [ARGS]...  Arguments for the command [default: None]                                     │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Options ──────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --volume  -v      source:target  Bind mount a volume, in Docker format: source:target.                             │
│ --gpus            'all' | int    IDs of host GPUs to make available in the tesseract. You can use all GPUs via     │
│                                  `--gpus all` or pass (multiple) IDs: `--gpus 0 --gpus 1`.                         │
│                                  [default: None]                                                                   │
│ --env     -e      key=value      Set environment variables in the Tesseract container, in Docker format:           │
│                                  key=value.                                                                        │
│ --user            TEXT           User to run the Tesseract as e.g. '1000' or '1000:1000' (uid:gid).                │
│                                  [default: None]                                                                   │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

- Now hidden from cli. 
- Rebased on top of Linus's #250 
- Added to serve (it wasn't there before oops!)
- Added unit test in serve 

@dionhaefner  I left the changes to @ because it makes sense if we have a mounted input directory that the @ filepaths would point to it... e.g.

```
 tesseract run --input-path examples/vectoradd vectoradd apply @example_inputs.json
```

Otherwise the interactions on the cli side with PAYLOAD wouldn't make much sense, even though you said you're okay with only passing the `--input-path` flag and doing nothing with it... It would also make it impossible to test until we have FileReference set up. And if we end up deciding to deprecate @ then this could all be removed.

angela-ko commented on tesseract_core/sdk/engine.py:
oh yeah definitely

angela-ko commented:
Reverted changes to `@infile` .... and all the unit testing that it comes with for a lighter and less intrusive PR. 

Now `--input-path` needs FileReference in order to be tested.

angela-ko commented on tesseract_core/runtime/cli.py:
```diff
@@ -347,6 +349,12 @@ def _create_user_defined_cli_command(
 
     options.extend(
         [
+            click.Option(
+                ["--input-path"],
+                type=click.STRING,
+                help="coming soon.",
+                default=None,
```
I don't think we need this -- I added this at some point before but realized there's no point in adding this since PAYLOAD gets parsed first before input-path is read, and for the runtime, they just need to set the env var.

angela-ko commented on tesseract_core/runtime/file_interactions.py:
```diff
@@ -14,6 +15,48 @@
 supported_format_type = Literal["json", "msgpack", "json+base64", "json+binref"]
 SUPPORTED_FORMATS = get_args(supported_format_type)
 
+os.environ["TESSERACT_CONTAINER_INPUT_PATH"] = "/tesseract/input_path"
```
Why do we need to differentiate between container input path and client input path instead of just the TESSERACT_INPUT_PATH. Once the `--input-path` is mounted, we don't really have a need to know the client's path as all we need to know is the mounted location which is set as `TESSERACT_INPUT_PATH` in my changes, in which case we also need to get the actual input path from the environment first. 

`TESSERACT_INPUT_PATH` would be the location that the client specified directory is mounted to in the case where we are running in a container and after mounting we do not need to know the original location anymore, and in the case where we are running runtime locally, the client would just set that directly as their local directory in the env var since no mounting would be happening. In both cases, we only need to know a single variable and not both.

angela-ko commented:
@nmheim Made some modifications based on the comment above (feel free to revert / edit). 

I think the only env var we need to know is `TESSERACT_INPUT_PATH` and `TESSERACT_OUTPUT_PATH`. We don't need to know what the path originally is as after it's bind mounted, we only need to know from the container side where the bind mount is located and we do not need to have any knowledge of where it is on the client's computer. If running runtime locally, then the user should just directly set where the input path is locally using `--input-path` or env var directly. Similarly with output path, we do not need to know where the client side's actual output path is for the container, but only where it got bind mounted to because that is what the container will be reading.

xalelax commented on examples/filereference/test_tesseract.py:
```diff
@@ -0,0 +1,40 @@
+
+# these are relative to input_path
+data = [
+    "sample_7.json",
+    "sample_6.json",
+    "sample_1.json",
+    "sample_0.json",
+    "sample_3.json",
+    "sample_2.json",
+    "sample_9.json",
+    "sample_5.json",
+    "sample_4.json",
+    "sample_8.json",
```
Why are they shuffled?

nmheim replied:
haha, no reason, I didn't even notice 😅

xalelax commented on tesseract_core/runtime/cli.py:
```diff
@@ -347,6 +349,16 @@ def _create_user_defined_cli_command(
+                help=(
+                    "Input path from which to read files. "
+                    "To be used in conjunction with InputFileReferences in the Tesseract InputSchema. "
+                    "See examples/filereference for a usage example."
```
What examples/filereference is will not be clear to someone who pip installed tesseract-core. I'd just remove this.
```suggestion
```

xalelax commented on tesseract_core/runtime/schema_types.py:
```diff
@@ -371,3 +373,27 @@ def is_differentiable(obj: Any) -> bool:
+    input_path = get_input_path()
+    tess_path = (input_path / path).resolve()
+    if str(input_path) not in str(tess_path):
+        raise ValueError("File references have to be relative to --input-path.")
```
I'd add something like "invalid path" or similar before the suggestion written here

xalelax commented on tesseract_core/sdk/engine.py:
```diff
@@ -603,6 +605,16 @@ def serve(
+            volumes = []
+        if "://" not in input_path:
+            input_path = Path(input_path).resolve()
+            volumes.append(f"{input_path}:/tesseract/input_data:ro")
```
This made me think: we should write somewhere (maybe docs) that a tesseract won't edit input files in-place

xalelax commented on tesseract_core/sdk/engine.py:
```diff
@@ -900,7 +914,18 @@ def run_tesseract(
+            volumes = []
+        if "://" not in input_path:
+            volumes.append(f"{input_path}:/tesseract/input_data")
+        # environment["TESSERACT_LOCAL_INPUT_PATH"] = str(input_path)
```
Dead code?
```suggestion
```

xalelax commented on tesseract_core/sdk/engine.py:
```diff
@@ -935,6 +960,8 @@ def run_tesseract(
 
             # Bind-mount directory
             parsed_volumes[str(local_path)] = {"bind": path_in_container, "mode": "rw"}
+            # environment["TESSERACT_LOCAL_OUTPUT_PATH"] = str(local_path)
```
Same?

nmheim commented on examples/filereference/tesseract_api.py:
```diff
@@ -0,0 +1,28 @@
+
+
+def apply(inputs: InputSchema) -> OutputSchema:
+    output_path = get_output_path()
```
you mean renaming `get_output_path` to `tesseract_output_path`? I like that!

nmheim commented on examples/filereference/tesseract_api.py:
```diff
@@ -0,0 +1,28 @@
+from pydantic import BaseModel
+
+from tesseract_core.runtime.file_interactions import get_output_path
+from tesseract_core.runtime.schema_types import InputFileReference, OutputFileReference
```
I moved it to `tesseract_core.runtime` like the other schema types, wdyt?

xalelax replied:
Personally I like stuff even less nested in Python, but good enough for now 🙂

nmheim commented on examples/filereference/README.md:
```diff
@@ -0,0 +1,26 @@
+
+Using `InputFileReference` and `OutputFileReference` you can
+include references to files in the `InputSchema` and `OuputSchema` of a Tesseract.
+The file reference schemas make sure that a file exists (either locally or in the Tesseract)
```
Yes, I think @llueg is waiting for this PR to be merged so that he can build on to of it 👍
